core/utils/codel.h

```diff
@@ -0,0 +1,331 @@
+                   uint64_t target, uint64_t window)
+    : delay_target_(target),
+      window_(window),
+      time_above_target_(0),
+      time_till_drop_(MicroSecondTime() + window),
+      drop_count_(0),
+      dropping_(0),
+      queue_(init_queue),
+      drop_func_(drop_func) {}
+
+template <template <typename type> class Q, typename T>
+Codel<Q, T>::~Codel() {
+  Wrapper* w;
+  int err = 0;
+  while ((w = queue_->Pop(&err))) {
+    Drop(w);
+  }
+}
+
+template <template <typename type> class Q, typename T>
+bool Codel<Q, T>::Full() {
+  return queue_->Full();
+}
+
+template <template <typename type> class Q, typename T>
+bool Codel<Q, T>::Empty() {
+  return queue_->Empty();
+}
+
+template <template <typename type> class Q, typename T>
+size_t Codel<Q, T>::Size() {
+  return queue_->Size();
+}
+
+template <template <typename type> class Q, typename T>
+size_t Codel<Q, T>::Capacity() {
+  return queue_->Capacity();
+}
+
+template <template <typename type> class Q, typename T>
+uint64_t Codel<Q, T>::MicroSecondTime() {
+  struct timeval tv;
+  gettimeofday(&tv, nullptr);
+  return tv.tv_sec * 1e6 + tv.tv_usec;
+}
+
+template <template <typename type> class Q, typename T>
+void Codel<Q, T>::Drop(Wrapper* w) {
+  if (w != nullptr) {
+    if (w->ptr != nullptr) {
+      drop_func_(w->ptr);
+    }
+    delete w;
+  }
+}
+
+template <template <typename type> class Q, typename T>
+uint64_t Codel<Q, T>::NextDrop(uint64_t cur_time) {
+  return cur_time + window_ * pow(drop_count_, -.5);
+}
+
+template <template <typename type> class Q, typename T>
+void Codel<Q, T>::Push(T* ptr, int* err) {
+  Wrapper* w = new Wrapper(ptr, MicroSecondTime());
+  queue_->Push(w, err);
+}
+
+template <template <typename type> class Q, typename T>
+void Codel<Q, T>::Push(T** ptr, size_t count, int* err) {
+  for (size_t i = 0; i < count; i++) {
+    Push(ptr[i], err);
+    if (*err != 0) {
+      break;
+    }
+  }
+}
+
+template <template <typename type> class Q, typename T>
+typename Codel<Q, T>::Wrapper* Codel<Q, T>::RingDequeue(int* err) {
+  Wrapper* w = queue_->Pop(err);
+  if (*err != 0) {
+    return nullptr;
+  }
+  uint64_t now = MicroSecondTime();
+  uint64_t delay_time = now - w->timestamp;
+
+  // determine whether object should be dropped or to change state
+  if (delay_time < delay_target_) {
+    time_above_target_ = 0;
+  } else {
+    if (time_above_target_ == 0) {
+      time_above_target_ = now + window_;
+    } else if (now >= time_above_target_) {
+      w->drop = 1;
+    }
+  }
+  return w;
+}
+
+template <template <typename type> class Q, typename T>
+typename Codel<Q, T>::Wrapper* Codel<Q, T>::DropDequeue(Wrapper* w, int* err) {
+  uint64_t now = MicroSecondTime();
+  if (!w->drop) {
+    dropping_ = 0;
+  } else if (now >= time_till_drop_) {
+    while (now >= time_till_drop_ && dropping_) {
+      Drop(w);
+      drop_count_++;
+      w = RingDequeue(err);
+      if (*err != 0) {
+        return nullptr;
```

Comment:
This may be a dumb question, but at this point, who owns w?  That is, if there was an error, does RingDequeue sometimes still return an object, and if so, who owns that memory / who is responsible for freeing it.  If RingDequeue never returns non-null upon error, then perhaps you can get rid of the err argument as suggested in an earlier comment.  (This applies below as well.)